bootloader/start64.c

```diff
@@ -67,16 +88,53 @@ uint64_t relocate_kernel (uint32_t mboot_struct_addr) {
   uint32_t bss_end_addr = kernel[i+6];
   uint32_t entry_addr = kernel[i+7];
 
+  /* Find the last used memory address */
+  uint64_t last_used = bss_end_addr;
+  if (last_used < (uint64_t)&mboot_struct[7]) {
```

Comment:
You should have a MAX function instead of lots of explicit ifs. This code is hard to read.